### PR TITLE
Python predict

### PR DESCRIPTION
**Pull request by ProskuraPD**
I add some tests for function python predict for single object

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=ru

### CODE REVIEW

File: catboost/python-package/catboost/core.py [hunk 1]
```diff
@@ -1215,18 +1215,30 @@ def fit(self, X, y=None, cat_features=None, pairs=None, sample_weight=None, grou
...
+            if isinstance(data, ARRAY_TYPES):
+                if isinstance(data[0], ARRAY_TYPES):
+                    data = Pool(
+                        data=data,
+                        cat_features=self._get_cat_feature_indices() if not isinstance(data, FeaturesData) else None
+                    )
+                else:
+                    data = [data]
+                    single_flag = True
+                    data = Pool(
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/catboost/core.py]: Please avoid copy-paste technique.

File: catboost/python-package/catboost/core.py [hunk 2]
```diff
@@ -1215,18 +1215,30 @@ def fit(self, X, y=None, cat_features=None, pairs=None, sample_weight=None, grou
...
+                        data=data,
+                        cat_features=self._get_cat_feature_indices() if not isinstance(data, FeaturesData) else None
+                    )
+            elif not isinstance(data, FeaturesData):
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/catboost/core.py]: Please revert the change in the elif branch.
Cat features passed to predict() must be a subset of cat features passed to fit().

File: catboost/python-package/ut/medium/run_catboost.py [hunk 3]
```diff
@@ -12,6 +12,7 @@ def test():
     model = CatBoostClassifier(iterations=5)
     model.fit(X, y, cat_features=categorical_features_indices)
     preds = model.predict(X)
+    pred_single = model.predict(X[0])
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/ut/medium/run_catboost.py]: Please add two new tests instead of modifying the existing tests

1. cat features, e.g. adult + fit() + predict(object) == predict([object])[0]
2. only float features, e.g. Higgs + fit() + predict(pool)[0] == predict(pool[0])

File: catboost/python-package/ut/medium/test.py [hunk 4]
```diff
@@ -2903,6 +2903,8 @@ def test_set_cat_features_in_init():
     label = np.random.randint(2, size=20)
     train_pool = Pool(data, label, cat_features=[1, 2])
     test_pool = Pool(data, label, cat_features=[1, 2])
+    #for single tests
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/ut/medium/test.py]: Please revert changes in this test, and add the tests described in the previous comment.

File: catboost/python-package/catboost/core.py [hunk 5]
```diff
@@ -1215,18 +1215,30 @@ def fit(self, X, y=None, cat_features=None, pairs=None, sample_weight=None, grou
                          save_snapshot, snapshot_file, snapshot_interval)
 
     def _predict(self, data, prediction_type, ntree_start, ntree_end, thread_count, verbose):
+        single_flag = False
```
Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/catboost/core.py]: Please update the description of the data parameter and the return value.

Comment (Evgueni-Petrov-aka-espetrov) [inline catboost/python-package/catboost/core.py]: For consistency, please apply the changes from this pull request to the following methods.

- CatBoost.staged_predict()
- CatBoostClassifier.predict()
- CatBoostClassifier.staged_predict()
- CatBoostClassifier.predict_proba()
- CatBoostClassifier.staged_predict_proba()
- CatBoostRegressor.predict()
- CatBoostRegressor.staged_predict()

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Polina, could you please resolve conflict in test.py?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 651837

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Could you please test as follows before committing?
'./ya make catboost/python-package -A'

**1. catboost/python-package/catboost (core.py::pep8)**
arcadia/catboost/python-package/catboost/core.py:1281:11: E271 multiple spaces after keyword
        if  is_data_single_object:

**2. catboost/python-package/ut/medium (test_whl.py::test_wheel)**

catboost/python-package/ut/medium/test_whl.py:24: in test_wheel
    yatest.common.execute([python_binary, test_script], env={'PYTHONPATH': os.path.join(os.getcwd(), 'catboost')})
library/python/testing/yatest_common/yatest/common/process.py:438: in execute
    res.wait(check_exit_code, timeout, on_timeout)
library/python/testing/yatest_common/yatest/common/process.py:304: in wait
    self._finalise(check_exit_code)
library/python/testing/yatest_common/yatest/common/process.py:311: in _finalise
    raise ExecutionError(self)
E   ExecutionError: Command 'https://proxy.sandbox.yandex-team.ru/782296760/build-release/catboost.tar/catboost/python-package/ut/medium/python_binary/catboost-python arcadia/catboost/python-package/ut/medium/run_catboost.py' has failed with code 1.
E   Errors:
E   ...redict(data, prediction_type, ntree_start, ntree_end, thread_count, verbose)
E     File "http://download-distbuild.n.yandex-team.ru/mds/103851/results_merger-11633137844262982650-0-4AC-854184/catboost/python-package/ut/medium/test-results/catboost-python-package-ut-medium/testing_out_stuff.tar/2/catboost/catboost/core.py", line 1261, in _predict
E       if not isinstance(data[0], ARRAY_TYPES):
E     File "contrib/python/pandas/pandas/core/frame.py", line 2059, in __getitem__
E       return self._getitem_column(key)
E     File "contrib/python/pandas/pandas/core/frame.py", line 2066, in _getitem_column
E       return self._get_item_cache(key)
E     File "contrib/python/pandas/pandas/core/generic.py", line 1386, in _get_item_cache
E       values = self._data.get(item)
E     File "contrib/python/pandas/pandas/core/internals.py", line 3541, in get
E       loc = self.items.get_loc(item)
E     File "contrib/python/pandas/pandas/indexes/base.py", line 2136, in get_loc
E       return self._engine.get_loc(self._maybe_cast_indexer(key))
E     File "contrib/python/pandas/pandas/index.pyx", line 139, in pandas.index.IndexEngine.get_loc
E     File "contrib/python/pandas/pandas/index.pyx", line 161, in pandas.index.IndexEngine.get_loc
E     File "contrib/python/pandas/pandas/hashtable_class_helper.pxi", line 732, in pandas.hashtable.PyObjectHashTable.get_item
E     File "contrib/python/pandas/pandas/hashtable_class_helper.pxi", line 740, in pandas.hashtable.PyObjectHashTable.get_item
E   KeyError: 0

**3. catboost/python-package/ut/medium/gpu (pytest)**
0 tests

In test_single_object_regressor: function uses no argument 'pool'

File: README.md [hunk 6]
```diff
@@ -63,6 +63,7 @@ Anna Veronika Dorogush, Andrey Gulin, Gleb Gusev, Nikita Kazeev, Liudmila Ostrou
 Anna Veronika Dorogush, Vasily Ershov, Andrey Gulin ["CatBoost: gradient boosting with categorical features support"](http://learningsys.org/nips17/assets/papers/paper_11.pdf). Workshop on ML Systems
 at NIPS 2017.
 
+Proskura Polina Dmitrievna
```
[the remark this answers is not in the document]
Comment (Evgueni-Petrov-aka-espetrov) [reply README.md]: Could you please remove the following text 'Proskura Polina Dmitrievna'?

File: catboost/python-package/ut/medium/test.py [hunk 7]
```diff
@@ -2050,6 +2051,52 @@ def test_get_metadata_notrain():
...
+def get_single_object_from_dataset(pool):
+    test_list = open(data_file(pool, 'test_small')).read().split('\n')[:-1]
+    for i in range(len(test_list)):
+        test_list[i] = test_list[i].split('\t')
+        for j in range(len(test_list[i])):
+            try:
+                test_list[i][j] = float(test_list[i][j])
+            except:
+                test_list[i][j] = test_list[i][j]
+    return random.choice(test_list)
```
Comment (yazevnul) [inline catboost/python-package/ut/medium/test.py]: Maybe instead of reading entire dataset and then selecting one out of it is better to first find out number of lines in a dataset, then choose random line out of it and then parse it?

Also, you are using `np.random.choice` which is not deterministic, and this will result in a flaky tests (basically test are going to be useless), so it's better to fix PRNG seed.

so it can be rewritten as:
```suggestion
     def _get_single_object_from_dataset(name, seed=20181215, rng=None):
         if rng is None:
             rng = np.random.RandomState(seed=seed)
             
         with open(data_file(pool, 'test_small') as f:
             lines = f.read().split('\n')
             
         line = rng.choice(lines)
         parsed = []
         for c in lines.split('\t'):
             try:
                 parsed.append(float(c))
             except:
                 parsed.append(c)
                 
         return parsed
```

File: catboost/python-package/ut/medium/test.py [hunk 8]
```diff
@@ -2050,6 +2051,52 @@ def test_get_metadata_notrain():
...
+
+
+@pytest.mark.parametrize('pool', ['higgs', 'adult'])
+def test_single_object_features():
+    train_pool = Pool(data_file(pool, 'train_small'), column_description=data_file(pool, 'train.cd'))
+    test_object = get_single_object_from_dataset(pool)
+    model = CatBoost({'iterations': 20, 'random_seed': 0, 'loss_function': 'RMSE', 'task_type': task_type, 'devices': '0'})
```
Comment (yazevnul) [inline catboost/python-package/ut/medium/test.py]: it's better to use a fixed training directory (`train_dir` parameter of `CatBoost` object c-tor), you can create one with `tempfile.mkdtemp`, our test frame work will make sure to proved proper `TEMPDIR` environment variable

otherwise since tests are running in parallel, different CatBoost training processes may interfere

File: catboost/python-package/ut/medium/test.py [hunk 9]
```diff
@@ -2050,6 +2051,52 @@ def test_get_metadata_notrain():
...
+    model.fit(train_pool)
+    assert model.predict(test_object) == model.predict([test_object])[0]
+    assert list(model.staged_predict(test_object)) == list(list(model.staged_predict([test_object]))[0])
+
+
+@pytest.mark.parametrize('pool', 'adult')
+def test_single_object_regressor():
+    train_pool = Pool(data_file(pool, 'train_small'), column_description=data_file(pool, 'train.cd'))
+    model = CatBoostRegressor(iterations=2, learning_rate=0.03, random_seed=0, task_type=task_type, devices='0')
```
Comment (yazevnul) [inline catboost/python-package/ut/medium/test.py]: there is no reason to set `devices` parameter, these are not GPU tests

File: catboost/python-package/ut/medium/test.py [hunk 10]
```diff
@@ -2050,6 +2051,52 @@ def test_get_metadata_notrain():
...
+    model.fit(train_pool)
+    test_object = get_single_object_from_dataset(pool)
+    assert model.predict(test_object) == model.predict([test_object])[0]
+    assert list(model.staged_predict(test_object)) == list(list(model.staged_predict([test_object]))[0])
+
+
+@pytest.mark.parametrize('pool', 'adult')
```
Comment (yazevnul) [inline catboost/python-package/ut/medium/test.py]: are you sure this compiles? I think you must provide `pool` as a parameter for the test, and you probably need to parametrize it as 
```suggestion
 @pytest.mark.parametrize('pool', ['adult'])
```

File: catboost/python-package/ut/medium/test.py (hunk 7 above)
Comment (yazevnul) [reply catboost/python-package/ut/medium/test.py]: but instead of wring yet another dataset parser you should probably use `catboost.Pool` instead

**Comment by yazevnul** [comment]
As a friendly advice, before submitting changes try running tests. You can do it in a following way:
```bash
cd catboost/python-package/ut/medium
# do not specify -j option, this way build system will be able to utilize all available cores of your machine
# -DHAVE_CUDA=no will disable build of GPU part of CatBoost (you changes are not related to GPU
# so it will save you A TONN of time, especially on a laptop)
ya make --build-release -DHAVE_CUDA=no
# actually run tests
ya make --build=release -ttt -DHAVE_CUDA=no
```

In general it all should be done by our CI (continuous integration), but we have some issues with CI for opensource version of our project, and it's still work in progress (to connect GitHub review process with CI for opensource version of CatBoost).

**Comment by annaveronika** [comment]
@ProskuraPD Do you plan to finish this?

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: 701291

**Comment by annaveronika** [comment]
Closing the pr since there is more than a month of inactivity. That's sad :(
